### ISSUE TITLE
LinterBuilder should accept tsconfig

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno_lint#762]
There are some options in tsconfig that we want to change default values for `deno` (https://github.com/denoland/deno/issues/11186). `deno_lint` should be able to take those options into account and warn about changes. 

We actually parse tsconfig into strictly type structure in `deno` so I'm not yet sure which approach would be better:
- create a custom rule in `deno` that will accept that strictly typed structure upon instantiation
- extract the parsing logic and types into separate crate and share with both `deno` and `deno_lint`

**Comment by magurotuna** [comment]
I've been thinking about the same kind of problem in #672.
In #672, we plan to use [`MediaType`] to determine whether the target file is TS or JS. The problem here is [`MediaType`] is defined in deno_cli so we can't pass the value of type [`MediaType`] to deno_lint directly.

Similarly, in this issue, we want to pass the typed tsconfig values (which are effectively [`EmitConfigOptions`] I guess) from deno_cli to deno_lint, but it's not doable since this type is defined in deno_cli.

If I had to pick up, I'd choose your second option ("extract the parsing logic ~"). But I think another option would be to define another type like `LintTsConfig` that is a subset of [`EmitConfigOptions`], and then implement `From<EmitConfigOptions>` for `LintTsConfig`. To put it concretely:

```rust
// in `deno_lint`

pub struct LintTsConfig {
  // define fields we need when linting
}

// and let `LinterBuilder` be able to take `LintTsConfig`
impl LinterBuilder {
  fn ts_config(mut self, ts_config: LintTsConfig) -> Self {
    self.ts_config = ts_config;
    self
  }
}
```

```rust
// in `deno_cli`

impl From<EmitConfigOptions> for deno_lint::LintTsConfig {
  fn from(x: EmitConfigOptions) -> Self { .. }
}

// pass `EmitConfigOptions` to `LinterBuilder`
let linter = LinterBuilder::default()
  .ts_config(emit_config_options.into())
  .build();
```

I'm in favor of this third option, because in this way we could avoid the crates dependency relations messing up and the same approach would be applicable to the [`MediaType`] issue I mentioned above.

[`MediaType`]: https://github.com/denoland/deno/blob/473713c6210ee11f11b7ae4c83165c4f87ff2d77/cli/media_type.rs#L16-L27
[`EmitConfigOptions`]: https://github.com/denoland/deno/blob/473713c6210ee11f11b7ae4c83165c4f87ff2d77/cli/config_file.rs#L23-L32

**Comment by bartlomieju** [comment]
@magurotuna I agree, it's better to duplicate a bit of code than introduce another inter-crate dependency. In linter we only care about 4 out of 9 variants in `MediaType` anyway (`JavaScript`, `Jsx`, `TypeScript`, `Dts`, `Tsx`). This should be way simpler to implement and leave most of complexity in CLI.

**Comment by kitsonk** [comment]
Two thoughts...

Is MediaType actually a core concept like ModuleSpecifier. While core doesn't use it, it has become a foundational concept where we want reliable handling. Should we move it to core?

Also while we are only concerned about a few keys in the compiler options part of the config at the moment, that structure is intended to be "the structure" for config, including deno lint and deno format. So again, maybe it is actually a core concept now?

**Comment by magurotuna** [comment]
@kitsonk 
Moving the two into deno_core would also be fine except that it leads to "fat" core. Also, as far as I understand deno_core is carefully kept to be TypeScript agnostic, right? I'm a bit worried that having TS-related stuff in deno_core would violate this philosophy.

**Comment by bartlomieju** [comment]
@kitsonk that sounds good to me, especially if you folks are already using this pattern and think this is useful.
